torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1553,21 +1580,11 @@ void ProcessGroupNCCL::watchdogHandler() {
               globalStore_->set(std::string(TIMEOUT_DUMP), vec);
             }
 
-            if (dumpOnTimeout_ && !optAsyncDebugDump) {
-              // Store debug info to storage. (By default to local disk)
-              optAsyncDebugDump = std::async(std::launch::async, [this]() {
-                return this->dumpDebuggingInfo();
-              });
-              // wait for the dump until timeout
-              waitForFutureOrTimeout(
-                  *optAsyncDebugDump,
-                  std::chrono::milliseconds(waitTimeoutDumpInMilSec_),
-                  "Flight recorder dump in watchdog");
-              // This sleep is used to give additional time for other ranks to
-              // receive the dump signal and then dump flight records, the same
-              // sleep time used in the heartbeatmonitor thread
-              // TODO, we should probably dump from a single thread in the
-              // future
+            if (dumpOnTimeout_) {
+              // signal the monitor thread to start dumping
+              shouldDump_.store(true);
+              // This sleep is used to give time for dumping before throwing
```

Comment:
i think we could also consider never throwing exceptions or commAborting from watchdog thread, and, using a similar means (perhaps storing a static exception ptr if we think of a better way) letting the monitor thread assume sole responsibility for 'exiting the process' if it's signaled to do so

could be a separate PR. wdyt?

Reply:
Yes, probably a single place to exit as well, will try it in another PR